src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/SemanticTokensEncoder.cs

```diff
@@ -0,0 +1,230 @@
+        /// If prev token and current token are on the same line, then start index is subtracted from prev token's start index else it remains same. </para>
+        /// 
+        /// <para>3) Third integer for each token is the length computed by subtracting token's end index and start index. </para>
+        /// 
+        /// <para>4) Fourth integer for each token is the token type encoded to an unsigned integer. 
+        /// Usually this is the position of token in the list or enum of all token types. </para>
+        /// 
+        /// <para>5) Fifth integer is a bit mask representing what token modifiers are applied to each token. this is always zero as powerfx doesn't have token modifiers.</para>
+        /// </summary>
+        /// <param name="tokens">Collection of tokens that would be encoded.</param>
+        /// <param name="expression">Expression from which these tokens were extracted from.</param>
+        /// <param name="eol">End of line character indiciating the line breaks in the given expression.</param>
+        /// <returns>Encoded Tokens.</returns>
+        // This encoding is done according to LSP specification for semantic tokens methods https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_semanticTokens
+        public static ICollection<uint> EncodeTokens(IEnumerable<ITokenTextSpan> tokens, string expression, string eol)
+        {
+            if (tokens == null)
+            {
+                return new uint[0];
+            }
+
+            var tokensByStartIdxs = new Dictionary<int, ITokenTextSpan>();
+            foreach (var token in tokens)
+            {
+                // Skip the tokens with a non-positive length
+                if (token.EndIndex - token.StartIndex <= 0)
+                {
+                    continue;
+                }
+
+                if (tokensByStartIdxs.TryGetValue(token.StartIndex, out var existingTokenAtCurrentTokenStartIdx))
+                {
+                    // Not likely to occur, but if multiple tokens start at the same index then pick the one with smallest end index
+                    // to potentially reduce the possibility of overlapping with tokens starting at different indexes
+                    // monaco-editor doesn't support overlapping tokens
+                    if (existingTokenAtCurrentTokenStartIdx.EndIndex > token.EndIndex)
+                    {
+                        tokensByStartIdxs[token.StartIndex] = token;
+                    }
+                }
+                else
+                {
+                    tokensByStartIdxs.Add(token.StartIndex, token);
+                }
+            }
+
+            if (tokensByStartIdxs.Count == 0)
+            {
+                return new uint[0];
+            }
+
+            ICollection<uint> encodedTokens = new LinkedList<uint>();
```

Comment:
Why not just a List<uint>? (and no need for the interface in the declaration, it's a local. Just use var)

Reply:
makes sense! i was debating between list vs linkedlist. Each token gets converted to 5 values and since we need to add 5 uints for  each token to the end of the list, i thought linked list would be better to avoid internal resizing of List<T> when its gets full. Also given how List<T>s are implemented and we need to do more insertions towards the end, we could end up with about twice (or whatever the growth factor is) the extra space. i don't have a strong opinion. In the end we just serialize whatever sequential collection we return to json array and send it back to the client.  We are not doing any other operations with the returned collection of uints.